curated_transformers/models/llama/decoder.py

```diff
@@ -0,0 +1,121 @@
+from typing import Any, List, Mapping, Optional, Type, TypeVar
+
+import torch
+from torch import Tensor
+from torch.nn import Dropout, Embedding, ModuleList
+
+from curated_transformers.models.normalization import RMSNorm
+
+from ..attention import AttentionMask
+from ..hf_hub import FromPretrainedHFModel
+from ..module import DecoderModule
+from ..output import KeyValueCache, ModelOutputWithCache
+from ._hf import convert_hf_config, convert_hf_state_dict
+from .config import LLaMAConfig
+from .layer import LLaMADecoderLayer
+
+# Only provided as typing.Self in Python 3.11+.
+Self = TypeVar("Self", bound="LLaMADecoder")
+
+
+class LLaMADecoder(DecoderModule, FromPretrainedHFModel):
+    """LLaMa (Touvron et al., 2023) decoder."""
+
+    def __init__(
+        self, config: LLaMAConfig, *, device: Optional[torch.device] = None
+    ) -> None:
+        """
+        :param config: Model configuration.
+        :param device: Device on which the module is to be initialized.
+        """
+        super().__init__()
+
+        self.embeddings = Embedding(
+            config.embedding.vocab_size, config.embedding.embedding_width, device=device
+        )
+        self.dropout = Dropout(p=config.embedding.dropout_prob)
+
+        self.layers = ModuleList(
+            [
+                LLaMADecoderLayer(config.layer, config.attention, device=device)
+                for _ in range(config.layer.num_hidden_layers)
+            ]
+        )
+
+        self.output_rms_norm = RMSNorm(
+            config.layer.hidden_width, eps=config.layer.rms_norm_eps, device=device
+        )
+
+    def forward(
+        self,
+        input_ids: Tensor,
+        attention_mask: Optional[AttentionMask] = None,
+        cache: Optional[List[KeyValueCache]] = None,
+        positions: Optional[Tensor] = None,
+        store_cache: bool = False,
+    ) -> ModelOutputWithCache[KeyValueCache]:
+        """
+        Apply the LLaMA decoder to the given piece identifiers.
+
+        :param input_ids: Piece identifiers to apply the decoder to.
+        :param attention_mask: Attention mask. Sequence elements for which the
+            corresponding mask element is set to ``False`` are ignored
+            during attention calculation.
+        :param cache: Key/value cache to avoid recomputing
+            key/value representations for tokens that were previously seen.
+        :param positions: Input positions. Positions are needed to
+            look up rotary embeddings. Normally, these positions are calculated
+            automatically. But if the positions deviate for some reason, they
+            can be provided through this argument.
+        :param store_cache: Whether to cache the key/value representations for
+            future reuse.
+        :returns: Decoder representations of the given pieces.
+
+        Shapes:
+            input_ids, attention_mask, positions - (batch, seq_len)
```

Comment:
Shapes as part of param docstrings.

Reply:
Fixed, also in NeoX, where this was originally copied from.